File: src/solar/solar-common/w3ts/handles/item.ts
/** @noSelfInFile **/

import {Handle} from "./handle";
import {MapPlayer} from "./player";
import {Point} from "./point";
import {Widget} from "./widget";
import DataBase from "../../common/DataBase";

export class Item extends Widget {

    // @ts-ignore
    public readonly handle!: item;

    constructor(itemId: number | string, x: number, y: number) {
        if (Handle.initFromHandle()) {
            super();
        } else {
            super(CreateItem(itemId, x, y));
        }
    }

    public get solarData(): AppItemData {
        return DataBase.getItemSolarData(this.handle)
    }

    public set solarData(obj: { [key: string]: any }) {
        DataBase.setDataByHandle("item", this.handle, obj)
    }

    /**
     * 物品的数量
     */
    public get charges() {
        return GetItemCharges(this.handle);
    }

    /**
     * 设置物品数量[使用次数]
     */
    public set charges(value: number) {
        SetItemCharges(this.handle, value);
    }

    /**
     * 设置物品是否无敌的
     */
    public set invulnerable(flag: boolean) {
        SetItemInvulnerable(this.handle, true);
    }

    /**
     * 物品是无敌的
     */
    public get invulnerable() {
        return IsItemInvulnerable(this.handle);
    }

    /**
     * 物品等级
     */
    public get level() {
        return GetItemLevel(this.handle);
    }

    /**
     * 物品名
     */
    get name() {
        return GetItemName(this.handle);
    }

    // set name(value: string) {
    //   BlzSetItemName(this.handle, value);
    // }
    /**
     * 物品可被抵押 [R]
     */
    public get pawnable() {
        return IsItemPawnable(this.handle);
    }

    /**
     * 设置物品能否变卖
     */
    public set pawnable(flag: boolean) {
        SetItemPawnable(this.handle, flag);
    }

    /**
     * 物品的所有者
     */
    public get player() {
        return GetItemPlayer(this.handle);
    }

    /**
     * 获取物品类型
     */
    public get type() {
        return GetItemType(this.handle);
    }

    /**
     * 物品的类别
     */
    public get typeId() {
        return GetItemTypeId(this.handle);
    }

    /**
     * 物品的类别
     */
    public get typeIdString() {
        return id2string(GetItemTypeId(this.handle));
    }

    /**
     * 物品自定义值
     */
    public get userData() {
        return GetItemUserData(this.handle);
    }

    /**
     * 设置物品自定义数据
     */
    public set userData(value: number) {
        SetItemUserData(this.handle, value);
    }

    /**
     * 物品可见 [R]
     */
    public get visible() {
        return IsItemVisible(this.handle);
    }

    /**
     * 显示/隐藏 [R]
     */
    public set visible(flag: boolean) {
        SetItemVisible(this.handle, flag);
    }

    // public get skin() {
    //   return BlzGetItemSkin(this.handle);
    // }
    //
    // public set skin(skinId: number | string) {
    //   BlzSetItemSkin(this.handle, skinId);
    // }
    /**
     * 物品的X轴坐标 [R]
     */
    public get x() {
        return GetItemX(this.handle);
    }

    /**
     * 设置物品坐标x [R]
     */
    public set x(value: number) {
        SetItemPosition(this.handle, value, this.y);
    }

    /**
     * 物品的Y轴坐标 [R]
     */
    public get y() {
        return GetItemY(this.handle);
    }

    /**
     * 设置物品坐标y [R]
     */
    public set y(value: number) {
        SetItemPosition(this.handle, this.x, value);
    }

    /**
     * 移除物品 [R]
     */
    public destroy() {
        DataBase.clearItemSolarData(this.handle)
        RemoveItem(this.handle);
    }

    // public getField(field: itembooleanfield | itemintegerfield | itemrealfield | itemstringfield) {
    //   const fieldType = field.toString().substr(0, field.toString().indexOf(":"));
    //
    //   switch (fieldType) {
    //     case "unitbooleanfield":
    //       return BlzGetItemBooleanField(this.handle, field as itembooleanfield);
    //     case "unitintegerfield":
    //       return BlzGetItemIntegerField(this.handle, field as itemintegerfield);
    //     case "unitrealfield":
    //       return BlzGetItemRealField(this.handle, field as itemrealfield);
    //     case "unitstringfield":
    //       return BlzGetItemStringField(this.handle, field as itemstringfield);
    //     default:
    //       return 0;
    //   }
    // }
    /**
     * 物品被持有 [R]
     */
    public isOwned() {
        return IsItemOwned(this.handle);
    }

    /**
     * 物品可被抵押 [R]
     */
    public isPawnable() {
        return IsItemPawnable(this.handle);
    }

    /**
     * 物品是拾取时自动使用的 [R]
     */
    public isPowerup() {
        return IsItemPowerup(this.handle);
    }

    /**
     * 物品可被市场随机出售 [R]
     */
    public isSellable() {
        return IsItemSellable(this.handle);
    }

    /**
     * 设置重生神符的产生单位类型
     */
    public setDropId(unitId: number | string) {
        SetItemDropID(this.handle, unitId);
    }

    /**
     * 设置物品是否死亡时掉落
     */
    public setDropOnDeath(flag: boolean) {
        SetItemDropOnDeath(this.handle, flag);
    }

    /**
     * 设置物品可以丢弃
     */
    public setDroppable(flag: boolean) {
        SetItemDroppable(this.handle, flag);
    }

    // public setField(field: itembooleanfield | itemintegerfield | itemrealfield | itemstringfield, value: boolean | number | string) {
    //   const fieldType = field.toString().substr(0, field.toString().indexOf(":"));
    //
    //   if (fieldType === "unitbooleanfield" && typeof value === "boolean") {
    //     return BlzSetItemBooleanField(this.handle, field as itembooleanfield, value);
    //   } else if (fieldType === "unitintegerfield" && typeof value === "number") {
    //     return BlzSetItemIntegerField(this.handle, field as itemintegerfield, value);
    //   } else if (fieldType === "unitrealfield" && typeof value === "number") {
    //     return BlzSetItemRealField(this.handle, field as itemrealfield, value);
    //   } else if (fieldType === "unitstringfield" && typeof value === "string") {
    //     return BlzSetItemStringField(this.handle, field as itemstringfield, value);
    //   }
    //
    //   return false;
    // }
    /**
     * 设置物品所属
     */
    public setOwner(whichPlayer: MapPlayer, changeColor: boolean) {
        SetItemPlayer(this.handle, whichPlayer.handle, changeColor);
    }

    /**
     * 移动物品到坐标(立即)(指定坐标) [R]
     */
    public setPoint(whichPoint: Point) {
        SetItemPosition(this.handle, whichPoint.x, whichPoint.y);
    }

    /**
     * 移动物品到坐标(立即)(指定坐标) [R]
     */
    public setPosition(x: number, y: number) {
        SetItemPosition(this.handle, x, y);
    }

    /**
     * 从handle转物品封装类
     */
    public static fromHandle(handle: item): Item {
        return this.getObject(handle);
    }

    /**
     * 物品可以抵押
     */
    public static isIdPawnable(itemId: number | string) {
        return IsItemIdPawnable(itemId);
    }

    /**
     * 物品可以充能
     */
    public static isIdPowerup(itemId: number | string) {
        return IsItemIdPowerup(itemId);
    }

    /**
     * 物品可以销售
     */
    public static isIdSellable(itemId: number | string) {
        return IsItemIdSellable(itemId);
    }

}

File: scripts/solar/solar-common/w3ts/handles/item.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["8"] = 3,["9"] = 3,["10"] = 6,["11"] = 6,["12"] = 7,["13"] = 7,["14"] = 9,["15"] = 9,["16"] = 9,["17"] = 9,["18"] = 14,["19"] = 15,["20"] = 9,["22"] = 9,["23"] = 18,["24"] = 18,["25"] = 18,["27"] = 14,["28"] = 183,["29"] = 184,["30"] = 185,["31"] = 183,["32"] = 207,["33"] = 208,["34"] = 207,["35"] = 214,["36"] = 215,["37"] = 214,["38"] = 221,["39"] = 222,["40"] = 221,["41"] = 228,["42"] = 229,["43"] = 228,["44"] = 235,["45"] = 236,["46"] = 235,["47"] = 242,["48"] = 243,["49"] = 242,["50"] = 249,["51"] = 250,["52"] = 249,["53"] = 271,["54"] = 272,["55"] = 271,["56"] = 278,["57"] = 279,["58"] = 278,["59"] = 285,["60"] = 286,["61"] = 285,["62"] = 292,["63"] = 293,["64"] = 292,["65"] = 299,["66"] = 300,["67"] = 299,["68"] = 306,["69"] = 307,["70"] = 306,["71"] = 313,["72"] = 314,["73"] = 313,["79"] = 23,["81"] = 26,["82"] = 27,["92"] = 34,["94"] = 40,["95"] = 41,["105"] = 55,["107"] = 47,["108"] = 48,["117"] = 62,["125"] = 69,["134"] = 79,["136"] = 85,["137"] = 86,["146"] = 93,["154"] = 100,["162"] = 107,["170"] = 114,["179"] = 121,["181"] = 127,["182"] = 128,["192"] = 135,["194"] = 141,["195"] = 142,["205"] = 156,["207"] = 162,["208"] = 163,["218"] = 170,["220"] = 176,["221"] = 177});
local ____exports = {}
local ____handle = require("solar.solar-common.w3ts.handles.handle")
local Handle = ____handle.Handle
local ____widget = require("solar.solar-common.w3ts.handles.widget")
local Widget = ____widget.Widget
local ____DataBase = require("solar.solar-common.common.DataBase")
local DataBase = ____DataBase.default
____exports.Item = __TS__Class()
local Item = ____exports.Item
Item.name = "Item"
__TS__ClassExtends(Item, Widget)
function Item.prototype.____constructor(self, itemId, x, y)
    if Handle:initFromHandle() then
        Widget.prototype.____constructor(self)
    else
        Widget.prototype.____constructor(
            self,
            CreateItem(itemId, x, y)
        )
    end
end
function Item.prototype.destroy(self)
    DataBase:clearItemSolarData(self.handle)
    RemoveItem(self.handle)
end
function Item.prototype.isOwned(self)
    return IsItemOwned(self.handle)
end
function Item.prototype.isPawnable(self)
    return IsItemPawnable(self.handle)
end
function Item.prototype.isPowerup(self)
    return IsItemPowerup(self.handle)
end
function Item.prototype.isSellable(self)
    return IsItemSellable(self.handle)
end
function Item.prototype.setDropId(self, unitId)
    SetItemDropID(self.handle, unitId)
end
function Item.prototype.setDropOnDeath(self, flag)
    SetItemDropOnDeath(self.handle, flag)
end
function Item.prototype.setDroppable(self, flag)
    SetItemDroppable(self.handle, flag)
end
function Item.prototype.setOwner(self, whichPlayer, changeColor)
    SetItemPlayer(self.handle, whichPlayer.handle, changeColor)
end
function Item.prototype.setPoint(self, whichPoint)
    SetItemPosition(self.handle, whichPoint.x, whichPoint.y)
end
function Item.prototype.setPosition(self, x, y)
    SetItemPosition(self.handle, x, y)
end
function Item.fromHandle(self, handle)
    return self:getObject(handle)
end
function Item.isIdPawnable(self, itemId)
    return IsItemIdPawnable(itemId)
end
function Item.isIdPowerup(self, itemId)
    return IsItemIdPowerup(itemId)
end
function Item.isIdSellable(self, itemId)
    return IsItemIdSellable(itemId)
end
__TS__SetDescriptor(
    Item.prototype,
    "solarData",
    {
        get = function(self)
            return DataBase:getItemSolarData(self.handle)
        end,
        set = function(self, obj)
            DataBase:setDataByHandle("item", self.handle, obj)
        end
    },
    true
)
__TS__SetDescriptor(
    Item.prototype,
    "charges",
    {
        get = function(self)
            return GetItemCharges(self.handle)
        end,
        set = function(self, value)
            SetItemCharges(self.handle, value)
        end
    },
    true
)
__TS__SetDescriptor(
    Item.prototype,
    "invulnerable",
    {
        get = function(self)
            return IsItemInvulnerable(self.handle)
        end,
        set = function(self, flag)
            SetItemInvulnerable(self.handle, true)
        end
    },
    true
)
__TS__SetDescriptor(
    Item.prototype,
    "level",
    {get = function(self)
        return GetItemLevel(self.handle)
    end},
    true
)
__TS__SetDescriptor(
    Item.prototype,
    "name",
    {get = function(self)
        return GetItemName(self.handle)
    end},
    true
)
__TS__SetDescriptor(
    Item.prototype,
    "pawnable",
    {
        get = function(self)
            return IsItemPawnable(self.handle)
        end,
        set = function(self, flag)
            SetItemPawnable(self.handle, flag)
        end
    },
    true
)
__TS__SetDescriptor(
    Item.prototype,
    "player",
    {get = function(self)
        return GetItemPlayer(self.handle)
    end},
    true
)
__TS__SetDescriptor(
    Item.prototype,
    "type",
    {get = function(self)
        return GetItemType(self.handle)
    end},
    true
)
__TS__SetDescriptor(
    Item.prototype,
    "typeId",
    {get = function(self)
        return GetItemTypeId(self.handle)
    end},
    true
)
__TS__SetDescriptor(
    Item.prototype,
    "typeIdString",
    {get = function(self)
        return id2string(GetItemTypeId(self.handle))
    end},
    true
)
__TS__SetDescriptor(
    Item.prototype,
    "userData",
    {
        get = function(self)
            return GetItemUserData(self.handle)
        end,
        set = function(self, value)
            SetItemUserData(self.handle, value)
        end
    },
    true
)
__TS__SetDescriptor(
    Item.prototype,
    "visible",
    {
        get = function(self)
            return IsItemVisible(self.handle)
        end,
        set = function(self, flag)
            SetItemVisible(self.handle, flag)
        end
    },
    true
)
__TS__SetDescriptor(
    Item.prototype,
    "x",
    {
        get = function(self)
            return GetItemX(self.handle)
        end,
        set = function(self, value)
            SetItemPosition(self.handle, value, self.y)
        end
    },
    true
)
__TS__SetDescriptor(
    Item.prototype,
    "y",
    {
        get = function(self)
            return GetItemY(self.handle)
        end,
        set = function(self, value)
            SetItemPosition(self.handle, self.x, value)
        end
    },
    true
)
return ____exports
